paddle/fluid/framework/fleet/index_wrapper.h

```diff
@@ -0,0 +1,176 @@
+    for (auto& ite : id_codes_map_) {
+      ids.push_back(ite.first);
+    }
+    return ids;
+  }
+
+  std::unordered_map<uint64_t, uint64_t> get_relation(
+      int level, const std::vector<uint64_t>& ids) {
+    std::unordered_map<uint64_t, uint64_t> pi_new;
+
+    for (auto& id : ids) {
+      auto code = id_codes_map_[id];
+      auto cur_level = meta_.height() - 1;
+      while (cur_level > level) {
+        code = (code - 1) / meta_.branch();
+        cur_level--;
+      }
+      pi_new[id] = code;
+    }
+    return pi_new;
+  }
+
+  std::vector<uint64_t> get_children_given_ancestor_and_level(
+      uint64_t ancestor, int level, bool ret_code = true) {
+    auto level_code_num =
+        static_cast<uint64_t>(std::pow(meta_.branch(), level));
+    auto code_min = level_code_num - 1;
+    auto code_max = level * level_code_num - 1;
+
+    std::vector<uint64_t> parent;
+    parent.push_back(ancestor);
+    std::vector<uint64_t> res;
+    size_t p_idx = 0;
+    while (true) {
+      size_t p_size = parent.size();
+      for (; p_idx < p_size; p_idx++) {
+        for (int i = 0; i < meta_.branch(); i++) {
+          auto code = parent[p_idx] * meta_.branch() + i + 1;
+          if (data_.find(code) != data_.end()) parent.push_back(code);
+        }
+      }
+      if ((code_min <= parent[p_idx]) && (parent[p_idx] < code_max)) {
+        break;
+      }
+    }
+
+    res = std::vector<uint64_t>(parent.begin() + p_idx, parent.end());
+    if (ret_code == false) {
+      for (size_t i = 0; i < res.size(); i++) {
+        res[i] = data_[res[i]].id();
+      }
+    }
+    return res;
+  }
+
+  std::vector<uint64_t> get_travel_path(uint64_t child, uint64_t ancestor) {
+    std::vector<uint64_t> res;
+    while (child > ancestor) {
+      res.push_back(data_[child].id());
+      child = (child - 1) / meta_.branch();
+    }
+    return res;
+  }
+
+  uint64_t tree_max_node() { return max_id_; }
+
+  int load(const std::string path);
+  std::unordered_map<uint64_t, Node> data_;
+  std::unordered_map<uint64_t, uint64_t> id_codes_map_;
+
+  uint64_t total_nodes_num_;
+  TreeMeta meta_;
+  uint64_t max_id_;
+};
+
+using TreePtr = std::shared_ptr<TreeIndex>;
+
+class IndexWrapper {
+ public:
+  virtual ~IndexWrapper() {}
+  IndexWrapper() {}
+
+  void clear_tree() { tree_map.clear(); }
+
+  TreePtr GetTreeIndex(const std::string name) {
+    PADDLE_ENFORCE_NE(tree_map.find(name), tree_map.end(),
+                      paddle::platform::errors::InvalidArgument(
+                          "tree [%s] doesn't exist. Please insert it firstly "
+                          "by API[\' insert_tree_index \'].",
+                          name));
+    return tree_map[name];
+  }
+
+  void insert_tree_index(const std::string name, const std::string tree_path) {
+    if (tree_map.find(name) != tree_map.end()) {
+      return;
+    }
+    TreePtr tree = std::make_shared<TreeIndex>();
+    int ret = tree->load(tree_path);
+    if (ret != 0) return;
```

Comment:
confused here.

Reply:
if ret != 0 means loading `tree_path` fails, nothing need to do.